pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,240 @@
+
+        // Update counters and let next element on the stack ignore the time we spent
+        final long delta = result.accumulate(timerEntry, extraDataCounter);
+        if (!queue.isEmpty()) { 
+            queue.peek().inNestedOperations += delta;
+        }
+    }
+    
+    private static class TimerEntry {
+        /* package */ final TimedOperation operation;
+        /* package */ final long start;
+        /* package */ long inNestedOperations = 0;
+        
+        /* package */ TimerEntry(final TimedOperationCategory category, final String label) {
+            this.operation = new TimedOperation(category, label);
+            this.start = System.nanoTime();
+        }
+
+        @Override
+        public String toString() {
+            return "TimerEntry for " + operation;
+        }
+    }
+    
+    /* package */ static class TimedResult {
+        /* package */ AtomicLong totalTime = new AtomicLong();
+        /* package */ AtomicLong selfTime = new AtomicLong();
+        /* package */ AtomicInteger callCount = new AtomicInteger();
+        /* package */ AtomicLong extraDataCounter = new AtomicLong();
+        
+        /* package */ long accumulate(final TimerEntry timerEntry, final long extraData) {
+            final long delta = System.nanoTime() - timerEntry.start;
+            
+            totalTime.getAndAdd(delta);
+            selfTime.getAndAdd(delta - timerEntry.inNestedOperations);
+            callCount.getAndIncrement();
+            extraDataCounter.getAndAdd(extraData);
+            
+            return delta;
+        }
+        
+        /* package */ void mergeTimes(final TimedResult timedResult) {
+            totalTime.getAndAdd(timedResult.totalTime.get());
+            selfTime.getAndAdd(timedResult.selfTime.get());
+        }
+    }
+    
+    /* package */ static class TimedOperation {
+        /* package */ final TimedOperationCategory category;
+        /* package */ final String label;
+        
+        /* package */ TimedOperation(final TimedOperationCategory category, final String label) {
+            this.category = category;
+            this.label = label;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + ((category == null) ? 0 : category.hashCode());
+            result = prime * result + ((label == null) ? 0 : label.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            TimedOperation other = (TimedOperation) obj;
+            if (category != other.category) {
+                return false;
+            }
+            if (label == null) {
```

Comment:
This `if` stmts and the following could be replaced with `return Objects.equals(label, other.label);`